src/frame/mod.rs

```diff
@@ -33,18 +35,31 @@ pub mod traits;
 
 use crate::error;
 
+static STREAM_ID: AtomicI16 = AtomicI16::new(0);
+
+pub type StreamId = i16;
+
 #[derive(Debug)]
 pub struct Frame {
     pub version: Version,
     pub flags: Vec<Flag>,
     pub opcode: Opcode,
-    pub stream: u16,
+    pub stream: StreamId,
     pub body: Vec<u8>,
     pub tracing_id: Option<Uuid>,
     pub warnings: Vec<String>,
 }
 
 impl Frame {
+    pub fn new(version: Version, flags: Vec<Flag>, opcode: Opcode, body: Vec<u8>, tracing_id: Option<Uuid>, warnings: Vec<String>) -> Self {
+        let mut stream = STREAM_ID.fetch_add(1, Ordering::SeqCst);
```

Comment:
Can we reset the stream id to 0 if it's reached the max value? After reaching it, each next `fetch_add` does not change the counter.

Reply:
`fetch_add` wraps around, and since negative ids are reserved, we simply add an offset.

Reply:
It must be some inconsistency in the spec:

https://github.com/apache/cassandra/blob/trunk/doc/native_protocol_v4.spec#L143 says that it can have negative values which are reserved for a server-inited streams.

https://github.com/apache/cassandra/blob/trunk/doc/native_protocol_v4.spec#L202 says that `[short]` is an unsigned integer. https://doc.rust-lang.org/std/primitive.u16.html

Anyway, let's maybe stay with your solution that eliminates negative values. 👍

Reply:
Given Java doesn't have unsigned primitives, I'd say it's signed.

Reply:
I understand. What I'm trying to say is that after reaching maximum for i16 each `fetch_add` won't have any effect. In this case it may make sense to reset the value to 0.

Reply:
I will, according to specs:

https://doc.rust-lang.org/std/sync/atomic/struct.AtomicI16.html#method.fetch_add

This operation wraps around on overflow.

Reply:
Could you please check https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a6650651cc307975f1fff491ce065d7b
It seems once we get a max, it won't be changed:

```
32767 -> 32767
```

Reply:
`fetch_add` returns the previous value, so you need to increment twice to see the overflow: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3b446f35b55187d7ac9838134f838e91

Reply:
Indeed. My bad. Thanks for your patience 🙂